native/cocos/renderer/gfx-metal/MTLDevice.mm

```diff
@@ -232,21 +234,21 @@ of this software and associated engine source code (the "Software"), a limited,
             [drawable release];
         }
         [cmdBuffer addCompletedHandler:^(id<MTLCommandBuffer> commandBuffer) {
-            onPresentCompleted();
+            onPresentCompleted(tempIndex);
         }];
         [cmdBuffer commit];
     }
 }
 
-void CCMTLDevice::onPresentCompleted() {
-    if (_currentBufferPoolId >= 0 && _currentBufferPoolId < MAX_FRAMES_IN_FLIGHT) {
-        CCMTLGPUStagingBufferPool *bufferPool = _gpuStagingBufferPools[_currentBufferPoolId];
+void CCMTLDevice::onPresentCompleted(uint32_t index) {
+    if (index >= 0 && index < MAX_FRAMES_IN_FLIGHT) {
+        CCMTLGPUStagingBufferPool *bufferPool = _gpuStagingBufferPools[index];
         if (bufferPool) {
             bufferPool->reset();
-            CCMTLGPUGarbageCollectionPool::getInstance()->clear(_currentBufferPoolId);
+            CCMTLGPUGarbageCollectionPool::getInstance()->clear(index);
         }
     }
-    _inFlightSemaphore->signal();
+    _inFlightSemaphores[index]->signal();
```

Comment:
whether index may exceed the size of the array

Reply:
index = ++index % 3 => index ranges {0, 2};
inFlightSemaphores.size() == 3. 
so it won't access out side of the array.